cmd/config_merger/main.go

```diff
@@ -0,0 +1,187 @@
+	wait     time.Duration
+}
+
+type configList struct {
+	Target  string    `json:"Target"`
+	Sources []sources `json:"Sources"`
+}
+
+type sources struct {
+	Name     string `json:"Name"`
+	Location string `json:"Location"`
+}
+
+func (o *options) getList() (*configList, error) {
+	if o.listPath == "" {
+		return nil, fmt.Errorf("'--config-list' required")
+	}
+	file, err := ioutil.ReadFile(o.listPath)
+	if err != nil {
+		return nil, err
+	}
+	foo := &configList{}
+	err = yaml.Unmarshal(file, foo)
+	if err != nil {
+		return nil, err
+	}
+	return foo, nil
+}
+
+func (o *options) validate() error {
+	if o.listPath == "" {
+		return errors.New("--config-list of configurations to merge required")
+	}
+	return nil
+}
+
+func gatherOptions() options {
+	var o options
+	flag.StringVar(&o.listPath, "config-list", "", "List of configurations to merge")
+	flag.StringVar(&o.creds, "gcp-service-account", "", "/path/to/gcp/creds (use local creds if empty)")
+	flag.BoolVar(&o.confirm, "confirm", false, "Upload data if set")
+	flag.DurationVar(&o.wait, "wait", 0, "Ensure at least this much time ahs passed since the last loop. (Run only once if zero)")
+	flag.Parse()
+	return o
+}
+
+func main() {
+	opt := gatherOptions()
+	err := opt.validate()
+	if err != nil {
+		logrus.WithError(err).Fatal("Invalid flags")
+	}
+	list, err := opt.getList()
+	if err != nil {
+		logrus.WithError(err).Fatalf("Could not read list at: %s", opt.listPath)
+	}
+	if !opt.confirm {
+		logrus.Info("--confirm=false (DRY-RUN): will not write to gcs")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client, err := gcs.ClientWithCreds(ctx, opt.creds)
+	if err != nil {
+		logrus.Fatalf("Failed to make storage client: %v", err)
+	}
+
+	target, err := gcs.NewPath(list.Target)
+	if err != nil {
+		logrus.WithError(err).Fatalf("Could not parse %s into a path", list.Target)
+	}
+
+	paths := make(map[string]*gcs.Path)
+	for _, source := range list.Sources {
+		path, err := gcs.NewPath(source.Location)
+		if err != nil {
+			logrus.WithError(err).Fatalf("Could not parse %s into a path", source.Location)
+			return
+		}
+		paths[source.Name] = path
+	}
+
+	if len(paths) == 0 {
+		logrus.Fatal("No shards to converge")
+		return
+	}
+
+	updateOnce := func(ctx context.Context) error {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+		defer cancel()
+		mergeAndUpdate(ctx, client, paths, *target, opt.confirm)
+		return nil
+	}
+
+	if err := updateOnce(ctx); err != nil {
+		logrus.WithError(err).Error("Failed update")
+	}
+	if opt.wait == 0 {
+		return
+	}
+	timer := time.NewTimer(opt.wait)
+	defer timer.Stop()
+	for range timer.C {
+		timer.Reset(opt.wait)
+		if err := updateOnce(ctx); err != nil {
+			logrus.WithError(err).Error("Failed update")
+		}
+		logrus.WithField("wait", opt.wait).Info("Sleeping")
+	}
+}
+
+func mergeAndUpdate(ctx context.Context, client *storage.Client, paths map[string]*gcs.Path, targetPath gcs.Path, confirm bool) {
```

Comment:
Will you just use a gcs.Client? It includes an uploader:

https://github.com/GoogleCloudPlatform/testgrid/blob/0180bbe30254944d8f3cdd1861e242d81afc6d76/util/gcs/client.go#L65-L66

Would also allow unit testing, which I think is a good idea.

Reply:
https://github.com/GoogleCloudPlatform/testgrid/blob/0180bbe30254944d8f3cdd1861e242d81afc6d76/pkg/updater/updater_test.go#L91-L245 an example that includes faking an upload as well as a fake config download

Reply:
Oh, this helps to explain the distinction between the `cmd/` and `pkg/` directories

I've done part of the refactor, and will move the mergeAndUpdate() function and write tests for it next.

Reply:
Refactor completed. The configuration and merge functions are in the `pkg/` directory with tests demonstrating when an upload is and isn't performed.